### ISSUE TITLE
Literal brackets in license text

### ISSUE THREAD

**Issue by dolmen** [github/choosealicense.com#542]
The text of the Apache 2.0 license on choosealicense.com has three lines different from the text at http://apache.org/licenses/LICENSE-2.0.txt

```diff
--- LICENSE.choosealicense.com	2017-10-19 00:33:45.000000000 +0200
+++ LICENSE.appache.org	2017-10-11 18:28:21.000000000 +0200
@@ -0,0 +1 @@
+
@@ -181 +182 @@
-      boilerplate notice, with the fields enclosed by brackets "{}"
+      boilerplate notice, with the fields enclosed by brackets "[]"
@@ -189 +190 @@
-   Copyright {yyyy} {name of copyright owner}
+   Copyright [yyyy] [name of copyright owner]
```

Note: The same problem appears also when initialising a new repo on GitHub, but I don't know where to report this issue.

**Comment by mlinksva** [comment]
Hi, I'm pretty sure this has been reported before, but I can't find a reference now. The reason is that the fields in the example notice in the trailer of Apache-2.0 (and *GPL) are not filled out in a deployed `LICENSE` file (they're intended to be used elsewhere, eg in source headers), but GitHub (which indirectly sources license texts from this repo) uses `[]` to fill in fields which are supposed to be replaced in a deployed `LICENSE` file (e.g., in MIT and other template licenses).

The difference you note ensures that GitHub won't modify the trailer material when people choose a license on GitHub. I agree it isn't ideal to modify license texts at all, but that's why, or my understanding anyway.

I'll look into whether there's anything that can be done to avoid the modification.

**Comment by mlinksva** [comment]
> (and *GPL)

Hmm, now that I look again, those use angle not square brackets in the original texts, so there's no good reason to substitute curly brackets. I'll make a PR shortly to fix.

The only other license cataloged here with the difference reported here is ECL-2.0, which is an Apache-2.0 derivative.

**Comment by wking** [comment]
> I'm pretty sure this has been reported before, but I can't find a reference now.

You may be thinking of #417 or #494.

**Comment by benbalter** [comment]
I believe https://github.com/github/choosealicense.com/issues/54 and https://github.com/github/choosealicense.com/pull/55 should explain the technical limitations at play (if any) here.

**Comment by wking** [comment]
Can we re-title this issue “Literal brackets in license text“ or some such?  It's not really Apache-specific.

And is resolving this issue really up to this repo?  I can't find GitHub's template-replacement logic here, and it may not be in a public repo at all or have public docs [besides these][0].  We could use the upstream licenses unchanged if the template-replacer whitelists the [known fields][1] and leaves other bracketed text alone.  Or we could escape literal brackets if the template-replacer has a syntax that allows that.  But without knowing anything about the tempate replacer, it's hard to know which of these would work.

[0]: https://github.com/github/choosealicense.com/blob/33f72ba7a77d7964914fcdc056dbd8f152f6b348/README.md#auto-populated-fields
[1]: https://github.com/github/choosealicense.com/blob/33f72ba7a77d7964914fcdc056dbd8f152f6b348/_data/fields.yml

**Comment by mlinksva** [comment]
@wking true, title changed. I'm going to see whether the technical limitations pointed out by @benbalter above are still at play. That will determine whether this issue will just be closed, or will be closed by a pull request to this repo. :smile:

**Comment by benbalter** [comment]
Did some digging into how the variables are interpolated within GitHub:

* For the actual rendering, we just use Mustache (meaning it looks for `{{{variable}}}`)
* Prior to rendering, we call `body.gsub("[", "{{{").gsub("]", "}}}")` to prepare the template for mustache
* We also have a process which uses `/(\[|\{\{\{)(?<field>.+?)(\]|\}\}\})/` to match fields (to prepare the HTML previews), but that process is aware of the potential fields and ignores any other value

So.. in terms of what's possible / easy to maintain, I'd suggest:

* License bodies in choosealicense.com should match the "official" text, as closely as possible, including brackets
* Since Licensee already knows the potential fields, we could add a `License#mustache_template` or similar method to convert known fields to a standardized format
* GitHub would simply rely on Licensee's open source implementation, making things easier to maintain in sync
* Any consumer of the license text would have to strictly interpolate only known fields

One gotcha: We can't use easily use`<`, as Markdown is a subset of HTML, when rendered in a Markdown file, or via GItHub Pages, it's treated as an HTML tag.

TL;DR: If anyone would like to submit a PR to normalize brackets, I'm glad to absorb the complexity upstream in Licensee.

**Comment by wking** [comment]
On Tue, Oct 24, 2017 at 03:58:45PM +0000, Ben Balter wrote:
> One gotcha: We can't use easily use`<`, as Markdown is a subset of
> HTML…

Can't we escape those in ‘License#mustache_template’ too?

**Comment by benbalter** [comment]
> Can't we escape those in ‘License#mustache_template’ too?

Yes, but:

1. When Jekyll renders the template it would render them as HTML tags
2. We'd be committing the escaped text to the repo, meaning it'll diverge from the known license text

**Comment by wking** [comment]
On Tue, Oct 24, 2017 at 09:25:46AM -0700, Ben Balter wrote:
> > Can't we escape those in ‘License#mustache_template’ too?
>
> Yes, but:
>
> 1. When Jekyll renders the template it would render them as HTML
>    tags
> 2. We'd be committing the escaped text to the repo, meaning it'll
>    diverge from the known license text

Maybe I'm not understanding something.  I was expecting we'd use the
upstream text (with raw <, etc.) in _licenses/*.txt here (just like we
do now [1]).  Anything that wants to consume raw text can continue to
consume it (however that works now).  Anything that wants Mustache
Markdown can consume mustache_template, which would insert {{{…}}} and
do other text → Markdown changes like ‘<’ → ‘&lt;’.  So I don't think
we need to commit &lt; in _licenses/*.txt, and I don't see how we'd be
breaking Jekyll.

[1]: https://github.com/github/choosealicense.com/blob/cca13baeb4574dcb4e8ef7a7a031369e8ce47e75/_licenses/agpl-3.0.txt#L40

**Comment by benbalter** [comment]
@wking I forgot we switched the licenses to .txt. You're aboslutely correct (and [we escape tags anyway](https://github.com/github/choosealicense.com/blob/gh-pages/_layouts/license.html#L48)) 👍

**Comment by benbalter** [comment]
> Since Licensee already knows the potential fields, we could add a License#mustache_template or similar method to convert known fields to a standardized format

https://github.com/benbalter/licensee/pull/231
